Action View: Document `TagBuilder` as part of the public interface

seanpdoyle opened the pull request:
### Motivation / Background

Since the `ActionView::Helpers::TagHelper::TagBuilder` class is marked with
`:nodoc:`, the `tag.attributes` method isn't directly discoverable by
the [API Documentation][], despite it being mentioned under the
[Building HTML Attributes][] subheading of the [tag][] helper method
documentation.

Similarly, methods like `tag.div` and `tag.h1` are also not
discoverable, since they're dependent on method missing.

### Detail

This commit promotes `ActionView::Helpers::TagHelper::TagBuilder` to be
part of Action View's public API. This commit marks all of its methods
as `:nodoc:` to omit them from the public interface, then adds RubyDoc
comments for `#attributes` and the generated tag building methods like
`#div` and `#a` to provide hooks for the [API Documentation][] to serve
search results.

To further reduce the publicly discoverable interface, this commit also
removes the `CaptureHelper` and `OutputSafetyHelper` modules from the
class. To preserve the existing behavior, delegate those methods to the
`@view_context` instance variable instead.

[API Documentation]: https://api.rubyonrails.org
[Building HTML Attributes]: https://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag-label-Building+HTML+attributes
[tag]: https://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

seanpdoyle committed:
Action View: Document `TagBuilder` as part of the public interface

Since the `ActionView::Helpers::TagHelper::TagBuilder` class is marked
with `:nodoc:`, the `tag.attributes` method isn't directly discoverable
by the [API Documentation][], despite it being mentioned under the
[Building HTML Attributes][] subheading of the [tag][] helper method
documentation.

Similarly, methods like `tag.div` and `tag.h1` are also not
discoverable, since they're dependent on method missing.

This commit promotes `ActionView::Helpers::TagHelper::TagBuilder` to be
part of Action View's public API. This commit marks all of its methods
as `:nodoc:` to omit them from the public interface, then adds RubyDoc
comments for `#attributes` and the generated tag building methods like
`#div` and `#a` to provide hooks for the [API Documentation][] to serve
search results.

To further reduce the publicly discoverable interface, this commit also
removes the `CaptureHelper` and `OutputSafetyHelper` modules from the
class. To preserve the existing behavior, delegate those methods to the
`@view_context` instance variable instead.

[API Documentation]: https://api.rubyonrails.org
[Building HTML Attributes]: https://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag-label-Building+HTML+attributes
[tag]: https://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag

zzak commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -53,11 +53,6 @@ def initialize(view_context)
-        # ERB.
-        #
-        #   <input <%= tag.attributes(type: :text, aria: { label: "Search" }) %> >
-        #   # => <input type="text" aria-label="Search">
```
I wonder if this method was public or meant to be public at one point? 🤔 

The other thought I have, it's perfectly fine and normal to have comments for internal use, as long as they are accurate. I've seen this a lot in AR, for example.

seanpdoyle replied:
It certainly is intended for public use. This commit originally replaced the comment with the `:method: tag.attributes` header comment in hopes of making the method more discoverable without duplicating the content.

I like your idea of retaining the original comment, though. I've added it back.

seanpdoyle commented:
@zzak I've attached some screenshots of the documentation pages:

**There isn't a documentation method for `attributes` defined on `TagBuilder`**:

![Screenshot 2023-09-25 at 12 52 16 PM](https://github.com/rails/rails/assets/2575027/a26ce740-591f-4ee1-b8f3-275d565b9d88)

There are now `tag.*` and `tag.attributes` entries:

![Screenshot 2023-09-25 at 12 52 25 PM](https://github.com/rails/rails/assets/2575027/a723ea6d-34f5-4008-b9ca-8831592b0226)

The `tag.attributes` method:

![Screenshot 2023-09-25 at 12 53 03 PM](https://github.com/rails/rails/assets/2575027/20dd9e11-90c0-4395-b8f4-03c6b6b91456)

The `tag.*` method:

![Screenshot 2023-09-25 at 12 53 20 PM](https://github.com/rails/rails/assets/2575027/d2f7f6fb-6ba4-4150-adcd-5bd42de9c564)

The table of contents:

![Screenshot 2023-09-25 at 12 53 40 PM](https://github.com/rails/rails/assets/2575027/5c307609-d8fa-4cb9-ba93-66abe2fcee00)

rafaelfranca commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -168,6 +168,112 @@ def method_missing(called, *args, **options, &block)
+      #   <button <%= tag.attributes id: "call-to-action", disabled: false, aria: { expanded: false } %> class="primary">Get Started!</button>
+      #   # => <button id="call-to-action" aria-expanded="false" class="primary">Get Started!</button>
+
+      # :method: tag.*
```
I don't think we use this kind of label to methods. This is going to be the first. Can't we document all the html tags?

seanpdoyle replied:
This style of documentation is derived from entries like [ActiveStorage::Attached::Model.with_attached_*](https://edgeapi.rubyonrails.org/classes/ActiveStorage/Attached/Model.html#method-i-with_attached_-2A).

The splat is supposed to serve as a searchable value, since both `TagBuilder` (the class returned by `helpers.tag`) is marked with `:nodoc:`, and because the methods themselves rely on `method_missing` and aren't defined.

rafaelfranca replied:
The difference is that we don't have any documentation that is `object.method`. The `with_attached_*` doesn't have an object that is a prefix. The right way to document this would be to expose the tag builder in the public API and perhaps stop using `method_missing`. That is probably why it is so slow compared with the `tag("input")` method.

seanpdoyle replied:
If we have access to Nokogiri, we could loop over `Nokogiri::HTML::ElementDescription::DefaultDescriptions`:

```ruby
irb(main):006:0> Nokogiri::HTML::ElementDescription::DefaultDescriptions.map(&:first).take(10)
=> ["a", "abbr", "acronym", "address", "applet", "area", "b", "base", "basefont", "bdo"]
```

rafaelfranca replied:
We could, but is there any reason to not hard code the list? We already have a list of `HTML_VOID_ELEMENTS` and `SVG_SELF_CLOSING_ELEMENTS`.

seanpdoyle commented:
@rafaelfranca @zzak I've incorporated your feedback into this diff. I've added RDoc methods for each of the generated element methods.

This PR description/title used to have `[ci skip]`, but no longer does. The full test suite should execute, since there are changes to the TagBuilder's inheritance.

zzak commented:
@seanpdoyle Could you try to `git commit --amend` and force push, since you've renamed the PR after the last push, I think.

seanpdoyle commented:
@zzak that worked! Thank you!

seanpdoyle commented:
@rafaelfranca @zzak are there other changes I could make to progress this work forward?

zzak commented:
I want to say I'm 👍 on this change, but declaring an interface (like this) as public has implications that require approval from core.
